stepflow-rs/crates/stepflow-cli/src/validate.rs

```diff
@@ -10,34 +10,58 @@
 // or implied. See the License for the specific language governing permissions and limitations under
 // the License.
 
-use crate::args::{ConfigArgs, load};
+use crate::args::{ConfigArgs, WorkflowLoader, load};
+use crate::schema_provider::PluginSchemaProvider;
 use crate::{MainError, Result, validation_display::display_diagnostics};
 use error_stack::ResultExt as _;
 use std::path::Path;
-use stepflow_analysis::validate_with_config;
+use stepflow_analysis::{type_check_to_diagnostics, validate_with_config};
+use stepflow_typecheck::{TypeCheckConfig, type_check_flow};
 
 /// Validate workflow files and configuration
 ///
 /// Returns the number of validation failures (errors + fatal diagnostics).
-pub async fn validate(flow_path: &Path, config_path: Option<&Path>) -> Result<u32> {
+pub async fn validate(
+    flow_path: &Path,
+    config_path: Option<&Path>,
+    type_check: bool,
+    strict: bool,
+) -> Result<u32> {
     // Load workflow and configuration
     let flow = load(flow_path)?;
+    let flow_dir = flow_path.parent();
 
     let config = if let Some(config_path) = config_path {
         // Load specific config file
         ConfigArgs {
             config_path: Some(config_path.to_path_buf()),
         }
-        .load_config(flow_path.parent())?
+        .load_config(flow_dir)?
     } else {
         // Auto-discover config
-        ConfigArgs { config_path: None }.load_config(flow_path.parent())?
+        ConfigArgs { config_path: None }.load_config(flow_dir)?
     };
 
     // Use unified validation
-    let diagnostics = validate_with_config(&flow, &config.plugins, &config.routing)
+    let mut diagnostics = validate_with_config(&flow, &config.plugins, &config.routing)
         .change_context(MainError::ValidationError("Validation failed".to_string()))?;
 
+    // Optionally run type checking
+    if type_check {
```

Comment:
Open an issue to create a more centralized cache for schema information (at least in the service). Perhaps validate should call a `validate` endpoint in the server?